### ISSUE TITLE
Leaking file descriptors in call to iptables

### ISSUE THREAD

**Issue by jasontibbitts** [denyhosts/denyhosts#33]
Basically, denyhosts calls os.system to shell out to iptables.  This leaks all of the open file descriptors to iptables, which causes selinux to have issues.  See the Fedora bug at https://bugzilla.redhat.com/show_bug.cgi?id=1197371

I think the best solution, given that denyhosts is actually security-sensitive, is to use a more secure way to call out to iptables.  At least `subprocess.Popen` with `close_fds=True`, but there is a small bit of code in firewalld we can copy.  (It's GPLv2+.)  In fact, I'll paste it here:

```python
def runProg(prog, argv=[ ]):
   args = [ prog ] + argv
    (rfd, wfd) = os.pipe()
    pid = os.fork()
    if pid == 0:
        try:
            fd = os.open("/dev/null", os.O_RDONLY)
            if fd != 0:
                os.dup2(fd, 0)
                os.close(fd)
            if wfd != 1:
                os.dup2(wfd, 1)
                os.close(wfd)
            os.dup2(1, 2)
            e = { "LANG": "C" }
            os.execve(args[0], args, e)
        finally:
            os._exit(255)
    os.close(wfd)

    cret = b''
    cout = os.read(rfd, 8192)
    while cout:
        cret += cout
        cout = os.read(rfd, 8192)
    os.close(rfd)
    (cpid, status) = os.waitpid(pid, 0)

    cret = cret.rstrip().decode('utf-8', 'replace')
    return (status, cret)
```

I think we could dump the output reading as well, though it might be nice to log it, I guess.

**Comment by mruffalo** [comment]
`subprocess.Popen(..., close_fds=True)` looks much nicer than manually juggling file descriptors between `fork` and `exec`, and it's supported as far back as 2.6. I'd vote for that instead of having a custom wrapper that reimplements part of the `subprocess` module. (Reimplements it a bit badly, I might add: that `argv` argument to `runProg` should not be a mutable list!)

**Comment by slicer69** [comment]
Jason, I agree with your point raised in the bug tracker that this should be corrected in DenyHosts. It looks to me like SELinux is giving a proper warning and we should address the issue.

As for the fix, I'm in favour of the shorter Popen call rather than pulling in a new function. I think the latter solution might be overkill.

**Comment by jasontibbitts** [comment]
Yes, I found out later that really just using Popen is the only reasonable way.  Some people like to plug they're "more super secure" code, and I'm still dumb enough to believe that it must somehow be better.

In any case, I'm just trying to put together the time to fix and test this.

Also, the selinux people just seem to have marked the thing 'dontaudit' which is terrible.  They love covering up legitimate bugs that way for some annoying reason, and if we leak FDs somewhere else we'll never know about it without semodule -DB and sifting through all of the cruft that generates because they dontaudit'ed way too many things without actually fixing them.

**Comment by slicer69** [comment]
Do you think we could ask them to revert their dontaudit patch once we fix our own code? Perhaps they can just rollback the change once we demostrate the warning has been addressed.

Marking something as dontaudit sounds too much like the "Just disable SELinux and try again" advice that was so common back around the time of Fedora Core 3.

**Comment by jasontibbitts** [comment]
I of course can ask, but they generally just do what they feel like doing.  They pushed an update even though I asked them not to do so.  Once we've fixed this I can file a bug asking for the rule to be removed.  Guess I need to learn enough about selinux to know where the rule actually is.
